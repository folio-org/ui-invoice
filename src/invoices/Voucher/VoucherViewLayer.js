import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import { FormattedMessage } from 'react-intl';
import { get } from 'lodash';

import {
  Button,
  Col,
  Icon,
  Layer,
  MenuSection,
  Pane,
  PaneMenu,
  Row,
} from '@folio/stripes/components';
import {
  AppIcon,
  stripesConnect,
} from '@folio/stripes/core';

import { LoadingPane } from '../../common/components';
import {
  invoiceResource,
  VOUCHER_BY_ID,
  VOUCHER_LINES,
} from '../../common/resources';
import VoucherView from './VoucherView';

const VoucherViewLayer = ({ match: { params }, mutator, resources }) => {
  const voucher = get(resources, ['voucher', 'records', 0]);
  const voucherLines = get(resources, ['voucherLines', 'records'], []);
  const vendorInvoiceNo = get(resources, ['invoice', 'records', 0, 'vendorInvoiceNo']);
  const isLoading = !get(resources, ['voucher', 'hasLoaded']);

  const closeVoucher = useCallback(
    () => {
      const _path = `/invoice/view/${params.id}`;

      mutator.query.update({ _path });
    },
    [params.id],
  );

  const renderActionMenu = () => (
    <MenuSection id="voucher-actions">
      <Button
        buttonStyle="dropdownItem"
        data-test-edit-voucher-button
      >
        <Icon size="small" icon="edit">
          <FormattedMessage id="ui-invoice.button.edit" />
        </Icon>
      </Button>
    </MenuSection>
  );

  const lastMenu = (
    <PaneMenu>
      <Button
        marginBottom0
        buttonStyle="primary"
      >
        <FormattedMessage id="ui-invoice.button.edit" />
      </Button>
    </PaneMenu>
  );

  if (isLoading) {
    return (
      <Layer isOpen>
        <LoadingPane onClose={closeVoucher} />
      </Layer>
    );
  }

  return (
    <Layer isOpen>
      <Pane
        actionMenu={renderActionMenu}
        appIcon={<AppIcon app="invoice" size="small" />}
        defaultWidth="fill"
        dismissible
        id="pane-voucher"
        lastMenu={lastMenu}
        onClose={closeVoucher}
        paneSub={
          <FormattedMessage
            id="ui-invoice.voucher.paneSubTitle"
            values={{ voucherNumber: voucher.voucherNumber }}
          />
        }
        paneTitle={
          <FormattedMessage
            id="ui-invoice.voucher.paneTitle"
            values={{ vendorInvoiceNo }}
          />
        }
      >
        <Row>
          <Col
            xs={12}
            md={8}
            mdOffset={2}
          >
            <VoucherView
              voucher={voucher}
              voucherLines={voucherLines}
            />
          </Col>
        </Row>
      </Pane>
    </Layer>
  );
};

VoucherViewLayer.manifest = Object.freeze({
  voucher: VOUCHER_BY_ID,
  voucherLines: {
    ...VOUCHER_LINES,
    params: {
      query: 'voucherId==:{voucherId}',
    },
  },
  invoice: invoiceResource,
  query: {},
});

VoucherViewLayer.propTypes = {
  match: ReactRouterPropTypes.match,
  mutator: PropTypes.object.isRequired,
  resources: PropTypes.object.isRequired,
};

export default stripesConnect(VoucherViewLayer);
