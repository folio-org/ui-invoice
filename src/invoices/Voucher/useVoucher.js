import { useQuery } from 'react-query';

import { useOkapiKy } from '@folio/stripes/core';

import {
  INVOICE_API,
  VOUCHERS_API,
  VOUCHER_LINES_API,
} from '../../common/constants';

export const useVoucher = (invoiceId, voucherId) => {
  const ky = useOkapiKy();

  const { isLoading: isVoucherLoading, data: voucher = {} } = useQuery(
    [VOUCHERS_API, voucherId],
    () => ky.get(`${VOUCHERS_API}/${voucherId}`).json(),
  );

  const { isLoading: isVoucherLinesLoading, data } = useQuery(
    [VOUCHER_LINES_API, voucherId],
    () => ky.get(`${VOUCHER_LINES_API}`, { query: `voucherId==${voucherId}` }).json(),
  );

  const { isLoading: isInvoiceLoading, data: invoice = {} } = useQuery(
    [INVOICE_API, invoiceId],
    () => ky.get(`${INVOICE_API}/${invoiceId}`, { query: `voucherId==${voucherId}` }).json(),
  );

  return ({
    isLoading: isVoucherLoading || isVoucherLinesLoading || isInvoiceLoading,
    voucher,
    voucherLines: data?.voucherLines || [],
    invoice,
  });
};
