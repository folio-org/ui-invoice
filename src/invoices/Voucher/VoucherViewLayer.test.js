import React from 'react';
import { render, act } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import {
  HasCommand,
  expandAllSections,
  collapseAllSections,
} from '@folio/stripes/components';

import {
  invoice,
  batchVoucher,
  batchVoucherLine,
  history,
  location,
  match,
} from '../../../test/jest/fixtures';

import { useVoucher } from './useVoucher';
import VoucherViewLayer from './VoucherViewLayer';

jest.mock('@folio/stripes-components/lib/Commander', () => ({
  HasCommand: jest.fn(({ children }) => <div>{children}</div>),
  expandAllSections: jest.fn(),
  collapseAllSections: jest.fn(),
}));

jest.mock('../PrintVoucher', () => ({ PrintVoucherContainer: jest.fn(() => 'PrintVoucherContainer') }));
jest.mock('./useVoucher', () => ({ useVoucher: jest.fn() }));
jest.mock('./VoucherView', () => {
  const { forwardRef } = jest.requireActual('react');

  return forwardRef(() => 'VoucherView');
});

const historyMock = {
  ...history,
  push: jest.fn(),
};
const matchParams = {
  id: 'id',
  voucherId: 'voucherId',
};
const defaultProps = {
  history: historyMock,
  location,
  match: {
    ...match,
    params: matchParams,
  },
};
const renderVoucherViewLayer = (props = defaultProps) => render(
  <VoucherViewLayer {...props} />,
  { wrapper: MemoryRouter },
);

describe('VoucherViewLayer', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;

    useVoucher.mockClear().mockReturnValue({
      isLoading: false,
      invoice,
      voucher: batchVoucher,
      voucherLines: [batchVoucherLine],
    });
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure', () => {
    const { asFragment } = renderVoucherViewLayer();

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Shortcuts', () => {
    beforeEach(() => {
      HasCommand.mockClear();
      expandAllSections.mockClear();
      collapseAllSections.mockClear();
    });

    it('should call expandAllSections when expandAllSections shortcut is called', async () => {
      renderVoucherViewLayer();

      act(() => {
        HasCommand.mock.calls[0][0].commands.find(c => c.name === 'expandAllSections').handler();
      });

      expect(expandAllSections).toHaveBeenCalled();
    });

    it('should call collapseAllSections when collapseAllSections shortcut is called', () => {
      renderVoucherViewLayer();

      act(() => {
        HasCommand.mock.calls[0][0].commands.find(c => c.name === 'collapseAllSections').handler();
      });

      expect(collapseAllSections).toHaveBeenCalled();
    });

    it('should navigate to voucher edit form when edit shortcut is called', () => {
      historyMock.push.mockClear();

      renderVoucherViewLayer();
      HasCommand.mock.calls[0][0].commands.find(c => c.name === 'edit').handler();

      expect(historyMock.push).toHaveBeenCalledWith({
        search: location.search,
        pathname: `/invoice/view/${matchParams.id}/voucher/${matchParams.voucherId}/edit`,
      });
    });

    it('should navigate to invoices list when search shortcut is called', () => {
      historyMock.push.mockClear();

      renderVoucherViewLayer();
      HasCommand.mock.calls[0][0].commands.find(c => c.name === 'search').handler();

      expect(historyMock.push).toHaveBeenCalledWith('/invoice');
    });
  });
});
