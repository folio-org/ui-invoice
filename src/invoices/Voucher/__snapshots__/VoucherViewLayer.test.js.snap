// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoucherViewLayer should render correct structure 1`] = `
<DocumentFragment>
  <div>
    <div
      class="paneset  "
    >
      <section
        class="pane focusIndicator"
        id="pane-voucher"
      >
        <div
          aria-labelledby="paneHeaderpane-voucher-pane-title"
          class="paneHeader"
          data-test-pane-header="true"
          id="paneHeaderpane-voucher"
          role="region"
          tabindex="-1"
        >
          <div
            class="paneHeaderButtonsArea first"
          >
            <div
              class="paneMenu"
            >
              <button
                aria-label="stripes-components.closeItem"
                class="iconButton medium paneHeaderCloseIcon PaneHeaderIconButton paneHeaderIconButton"
                data-test-pane-header-dismiss-button="true"
                icon="times"
                type="button"
              >
                <span
                  class="iconButtonInner mediumOffset PaneHeaderIconButton__inner"
                  data-test-pane-header-dismiss-button="true"
                >
                  <span>
                    Icon
                  </span>
                </span>
              </button>
            </div>
          </div>
          <div
            class="paneHeaderCenter"
          >
            <div
              class="paneHeaderCenterInner"
              id="paneHeaderpane-voucher-pane-title"
            >
              <h2
                class="paneTitle"
                data-test-pane-header-title="true"
              >
                <div
                  class="appIcon small icon-alignment-center"
                >
                  <span
                    aria-hidden="true"
                    class="icon"
                  />
                </div>
                <span
                  class="paneTitleLabel"
                >
                  ui-invoice.voucher.paneTitle
                </span>
              </h2>
              <p
                class="paneSub"
                data-test-pane-header-sub="true"
                id="paneHeaderpane-voucher-subtitle"
              >
                <span>
                  ui-invoice.voucher.paneSubTitle
                </span>
              </p>
            </div>
          </div>
          <div
            class="paneHeaderButtonsArea last"
          >
            <div
              class="paneMenu"
            >
              <div
                class="dropdown"
                data-pane-header-actions-dropdown="true"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="button primary marginBottom0 actionMenuToggle"
                  data-test-pane-header-actions-button="true"
                  type="button"
                >
                  <span
                    class="inner"
                  >
                    <span>
                      Icon
                    </span>
                  </span>
                </button>
                <div
                  style="display: none;"
                >
                  <div
                    class="DropdownMenu"
                    style="left: 0px; display: block;"
                  >
                    <section
                      aria-labelledby="voucher-actions-label"
                      class="menuSection"
                      data-test-menu-section="true"
                      id="voucher-actions"
                    >
                      <a
                        class="button dropdownItem"
                        data-test-edit-voucher-button="true"
                        href="/invoice/view/id/voucher/voucherId/edit?search"
                        role="button"
                      >
                        <span>
                          Icon
                        </span>
                      </a>
                      <button
                        class="button dropdownItem"
                        type="button"
                      >
                        <span
                          class="inner"
                        >
                          <span>
                            Icon
                          </span>
                        </span>
                      </button>
                    </section>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="paneContent hasPadding"
          id="pane-voucher-content"
          tabindex="-1"
        >
          <div
            class="row"
          >
            <div
              class="col-xs-12 col-md-8 col-md-offset-2"
            >
              VoucherView
            </div>
          </div>
        </div>
      </section>
    </div>
    <div
      class="container"
    >
      <div
        class="cursorHover"
        hidden=""
        style="left: -10px;"
      >
        <div
          class="cursor"
        >
          Â 
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
